crates/spark/src/tree/select_helper.rs

```diff
@@ -0,0 +1,194 @@
+use tracing::trace;
+
+use crate::tree::{
+    LeavesReservation, TargetAmounts, TargetLeaves, TreeNode, TreeService, TreeServiceError,
+};
+
+pub fn select_leaves_by_amounts(
+    leaves: &[TreeNode],
+    target_amounts: Option<&TargetAmounts>,
+) -> Result<TargetLeaves, TreeServiceError> {
+    let mut remaining_leaves = leaves.to_vec();
+
+    // If no target amounts are specified, return all remaining leaves
+    let Some(target_amounts) = target_amounts else {
+        trace!("No target amounts specified, returning all remaining leaves");
+        return Ok(TargetLeaves::new(remaining_leaves, None));
+    };
+
+    // Select leaves that match the target amount_sats
+    let amount_leaves = select_leaves_by_amount(&remaining_leaves, target_amounts.amount_sats)?
+        .ok_or(TreeServiceError::UnselectableAmount)?;
+
+    let fee_leaves = match target_amounts.fee_sats {
+        Some(fee_sats) => {
+            // Remove the amount_leaves from remaining_leaves to avoid double spending
+            remaining_leaves.retain(|leaf| {
+                !amount_leaves
+                    .iter()
+                    .any(|amount_leaf| amount_leaf.id == leaf.id)
+            });
+            // Select leaves that match the fee_sats from the remaining leaves
+            Some(
+                select_leaves_by_amount(&remaining_leaves, fee_sats)?
+                    .ok_or(TreeServiceError::UnselectableAmount)?,
+            )
+        }
+        None => None,
+    };
+
+    Ok(TargetLeaves::new(amount_leaves, fee_leaves))
+}
+
+/// Selects leaves from the tree that sum up to exactly the target amount.
+/// If such a combination of leaves does not exist, it returns `None`.
+pub(crate) fn select_leaves_by_amount(
```

Comment:
These `pub(crate)` helper functions could also potentially be reused outside of the crate?